src/codegen/compile.ml

```diff
@@ -3852,6 +3895,118 @@ module Blob = struct
+  let logical_grow env =
+    compile_const_64 (Int64.of_int (!Flags.max_stable_pages)) ^^
+    E.call_import env "rts" "region0_grow"
+
+  let load_blob env = E.call_import env "rts" "region0_load_blob"
+  let store_blob env = E.call_import env "rts" "region0_store_blob"
+
+  let load_word8 env = E.call_import env "rts" "region0_load_word8"
+  let store_word8 env = E.call_import env "rts" "region0_store_word8"
+
+  let load_word16 env = E.call_import env "rts" "region0_load_word16"
+  let store_word16 env = E.call_import env "rts" "region0_store_word16"
+
+  let load_word32 env = E.call_import env "rts" "region0_load_word32"
+  let store_word32 env = E.call_import env "rts" "region0_store_word32"
+
+  let load_word64 env = E.call_import env "rts" "region0_load_word64"
+  let store_word64 env = E.call_import env "rts" "region0_store_word64"
+
+  let load_float64 env = E.call_import env "rts" "region0_load_float64"
+  let store_float64 env = E.call_import env "rts" "region0_store_float64"
+end
+
+module Region = struct
+  (*
+    See rts/motoko-rts/src/region.rs
+   *)
+
+  (* Object layout:
+
+     ┌─────┬──────────┬──────────────────┬─────────────────┐
+     │ tag │ id_field │ page_count_field │ vec_pages_field │
+     └─────┴──────────┴──────────────────┴─────────────────┘
+            (unboxed, low 16 bits, rest 0-initialized padding)
+                        unboxed u32
+                                          Blob
+  *)
+
+  let id_field env = Int32.add (Tagged.header_size env) 0l
+  let page_count_field env = Int32.add (Tagged.header_size env) 1l
+  let vec_pages_field env = Int32.add (Tagged.header_size env) 2l
+
+  let load_field env region_field =
+    let field = region_field env in
+    assert (field = id_field env || field = page_count_field env || field = vec_pages_field env);
+    Tagged.load_forwarding_pointer env ^^
+    Tagged.load_field env field
+
+  let store_field env region_field =
+    let field = region_field env in
+    assert (field = id_field env || field = page_count_field env || field = vec_pages_field env);
+    let (set_value, get_value) = new_local env "value" in
+    set_value ^^
+    Tagged.load_forwarding_pointer env ^^
+    get_value ^^
+    Tagged.store_field env field
+
+  let alloc env get_id get_pagecount get_vec_pages =
+    Tagged.obj env Tagged.Region [ get_id; get_pagecount; get_vec_pages ]
+
+  let sanity_check s env =
+    if !Flags.sanity then
+    Func.share_code1 env ("check_region_" ^ s) ("val", I32Type) [I32Type]
+      (fun env get_region ->
+         get_region ^^ Tagged.load_tag env ^^
+         compile_eq_const Tagged.(int_of_tag Region) ^^
+         get_region ^^ Tagged.load_tag env ^^ (* repeated, optimize? *)
+         compile_eq_const Tagged.(int_of_tag StableSeen) ^^
+         G.i (Binary (Wasm.Values.I32 I32Op.Or)) ^^
+         E.else_trap_with env ("Internal error: bad region tag "^ s) ^^
+         get_region ^^ Tagged.load_field env (vec_pages_field env) ^^ Tagged.load_tag env ^^
```

Comment:
It is important to use the `Region.load_field` and not `Tagged.load_field` to go through the forwarding pointer:

```suggestion
         get_region ^^ load_field env (vec_pages_field env) ^^ Tagged.load_tag env ^^
```

Reply:
My bad.

Reply:
Maybe, it could make sense to define explicit access functions for each of the fields: `load_id_field`, `store_id_field`, etc. and remove the generic `load_field`/`store_field` in the `Region` module because it can be easily confused with `Tagged.load_field` and `Tagged.store_field`.

Reply:
Moreover, `store_vec_pages` is potentially dangerous as it may involve a write barrier. The usages here are initializations which are again okay. So, perhaps, having an initialization function for all three fields would be more robust and provide three different load functions for each of the region fields.